Add accessors for headers

Fryguy opened the pull request:
This PR adds accessors for specific header keys in the Request, and by doing so separates the "header" knowledge from the User class.  That is, the user just gets the identity hash from the request and doesn't have to know it's a header.

@syncrou Please review

Built on #43 so just review the second commit.
Closes #36

bzwei commented on lib/manageiq/api/common/request.rb:
```diff
@@ -31,17 +44,24 @@ def self.with_request(request)
         attr_reader :headers, :original_url
 
-        def initialize(headers:, original_url:, **kwargs)
+        def initialize(headers:, original_url:, **_kwargs)
```
should we remove the unused `**_kwargs`?

Fryguy replied:
This is there to capture and "extra" parameters that might be sent along, otherwise the caller blows up.  We can probably remove it if we are sure all callers are doing the right thing.  I'd leave that to a separate PR, though.

bzwei commented on lib/manageiq/api/common/request.rb:
```diff
@@ -31,17 +44,24 @@ def self.with_request(request)
+        end
+
+        def identity
+          JSON.parse(Base64.decode64(headers[IDENTITY_KEY]))
```
why not cached the identity here rather than in each user instance?
should raise an error if headers does not have key `IDENTITY_KEY`.

Fryguy replied:
I like the idea of caching it here.  I'll do that.

If the identity key is missing, should this raise or just return nil.  I was thinking just return nil, but I'm not sure.

syncrou replied:
From an RBAC perspective if the user doesn't exist nothing can happen.  So maybe we should `raise` instead of return `nil`.

bzwei commented on lib/manageiq/api/common/request.rb:
```diff
@@ -31,17 +44,24 @@ def self.with_request(request)
+        end
+
         def user
           @user ||= User.new
```
Can we return `nil` if header does not have `IDENTITY_KEY`? The other option is to raise an error.

Fryguy replied:
I agree this code should do that as well (or perhaps User.new should do that and it would bubble up to here).

syncrou commented on lib/manageiq/api/common/user.rb:
```diff
@@ -1,9 +1,9 @@
     module Common
-      class User
-        IDENTITY_KEY = 'x-rh-identity'.freeze
+      class IdentityError < StandardError; end
```
Same question about errors

Fryguy replied:
Yup see https://github.com/ManageIQ/manageiq-api-common/pull/44#discussion_r263913837 ☝️

Fryguy commented on lib/manageiq/api/common/request.rb:
```diff
@@ -1,13 +1,26 @@
   module API
     module Common
+      class RequestNotSet < ArgumentError
+        def initialize
```
only that it is directly related to the request object.  I really wanted to move it *inside* the Request class (i.e. `ManageIQ::API::Common::Request::NotSet`), but this was more of a first step.

Fryguy commented:
Updated with the suggestions.  Also, based on the rationale that the User class can be isolated form the "headers", I've changed the signature of User.new to take an identity payload (not headers, just the guts of the identity payload).  This way we can also test it in isolation from the headers themselves.  I did not yet change the specs for user though to not use headers...I can do that in a follow up PR.
